src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp

```diff
@@ -4,161 +4,219 @@
 
     m_hash = get_hash();
 }
 
 bool LazyTensorImpl::operator==(const LazyTensorImpl& other) const {
     if (m_hash != other.m_hash || m_orig_data != other.m_orig_data || m_orig_shape != other.m_orig_shape ||
-        m_orig_type != other.m_orig_type || m_transform.first != other.m_transform.first) {
+        m_orig_type != other.m_orig_type) {
         return false;
     }
 
-    switch (m_transform.first) {
-    case TransformType::THIS:
-        // everything is already compared above - skip
-        break;
-    case TransformType::CONVERT:
-        // everything is already compared above - skip
-        break;
-    case TransformType::PERMUTE:
-        if (std::get<std::vector<std::size_t>>(m_transform.second) !=
-            std::get<std::vector<std::size_t>>(other.m_transform.second)) {
-            return false;
-        }
-        break;
-    case TransformType::CONCAT:
-        if (std::get<ConcatMeta>(m_transform.second) != std::get<ConcatMeta>(other.m_transform.second)) {
-            return false;
-        }
-        break;
-    case TransformType::UNPACK:
-        if (std::get<UnpackMeta>(m_transform.second) != std::get<UnpackMeta>(other.m_transform.second)) {
-            return false;
-        }
-        break;
-    default:
-        NPUW_ASSERT(false);
-        break;
-    }
-
-    if ((m_parent && !other.m_parent) || (!m_parent && other.m_parent)) {
+    bool meta_diff = false;
+    std::visit(overloaded{[](const auto& op) { /* do nothing */ },
+                          [&](const op::Permute& op) {
+                              meta_diff = (op.axes != std::get<op::Permute>(other.m_transform).axes ||
+                                           op.tensor != std::get<op::Permute>(other.m_transform).tensor);
+                          },
+                          [&](const op::Convert& op) {
+                              meta_diff = (op.type != std::get<op::Convert>(other.m_transform).type ||
+                                           op.tensor != std::get<op::Convert>(other.m_transform).tensor);
+                          },
+                          [&](const op::Concat& op) {
+                              meta_diff = (op.axis != std::get<op::Concat>(other.m_transform).axis ||
+                                           op.tensors != std::get<op::Concat>(other.m_transform).tensors);
+                          },
+                          [&](const op::Unpack& op) {
+                              meta_diff = (op.type != std::get<op::Unpack>(other.m_transform).type ||
+                                           op.shape != std::get<op::Unpack>(other.m_transform).shape ||
+                                           op.w != std::get<op::Unpack>(other.m_transform).w ||
+                                           op.z != std::get<op::Unpack>(other.m_transform).z ||
+                                           op.s != std::get<op::Unpack>(other.m_transform).s);
+                          }},
+               m_transform);
```

Comment:
you'd better introduce `operator==` in your `op::` types and dispatch here to `T::operator==`. Let's type define the behavior, not the scattered code we have around

Reply:
Done